pkg/querier/remote/read.go

```diff
@@ -0,0 +1,187 @@
+package remote
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/config"
+
+	"github.com/grafana/loki/pkg/iter"
+	"github.com/grafana/loki/pkg/logcli/client"
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/logql"
+	"github.com/grafana/loki/pkg/logqlmodel"
+	"github.com/grafana/loki/pkg/querier"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
+	"github.com/grafana/loki/pkg/storage/stores/index/stats"
+)
+
+// ReadConfig is the configuration for reading from remote storage.
+// code from /github.com/prometheus/prometheus/config/config.go:868
+type ReadConfig struct {
+	URL   *config.URL `yaml:"url"`
+	Name  string      `yaml:"name,omitempty"`
+	OrgID string      `yaml:"orgID,omitempty"`
+
+	HTTPClientConfig config.HTTPClientConfig `yaml:",inline"`
+
+	//todo:@liguozhong support http herder
+	Headers map[string]string `yaml:"headers,omitempty"`
+	//todo:@liguozhong support RemoteTimeout
+	RemoteTimeout time.Duration `yaml:"remote_timeout,omitempty"`
+}
+
+func NewQuerier(name string, remoteReadConfig ReadConfig) (querier.Querier, error) {
+	client := &client.DefaultClient{
+		OrgID:   remoteReadConfig.OrgID,
+		Address: remoteReadConfig.URL.String(),
+		TLSConfig: config.TLSConfig{
+			InsecureSkipVerify: remoteReadConfig.HTTPClientConfig.TLSConfig.InsecureSkipVerify,
+		},
+	}
+	if remoteReadConfig.HTTPClientConfig.BasicAuth != nil {
+		client.Username = remoteReadConfig.HTTPClientConfig.BasicAuth.Username
+		client.Password = strings.TrimSpace(string(remoteReadConfig.HTTPClientConfig.BasicAuth.Password))
+	}
+
+	return &Querier{
+		client: client,
+		name:   name,
+	}, nil
+}
+
+type Querier struct {
+	client client.Client
+	codec  queryrangebase.Codec
+	name   string
+}
+
+func (q Querier) SelectLogs(ctx context.Context, params logql.SelectLogParams) (iter.EntryIterator, error) {
+	response, err := q.client.QueryRange(ctx, params.Selector, int(params.Limit), params.Start, params.End, params.Direction, 0, 0, false)
```

Comment:
So we can mimic streaming a little. The logcli client already uses batches to fetch results. We could change it so that it returns an iterator over the entries. This way we won't fetch too much all at once.

Reply:
return iter.NewStreamsIterator(streams.ToProto(), params.Direction), nil

pkg/iter/entry_iterator.go:452

// NewStreamsIterator returns an iterator over logproto.Stream
func NewStreamsIterator(streams []logproto.Stream, direction logproto.Direction) EntryIterator {
	is := make([]EntryIterator, 0, len(streams))
	for i := range streams {
		is = append(is, NewStreamIterator(streams[i]))
	}
	return NewSortEntryIterator(is, direction)
}

Reply:
> So we can mimic streaming a little. The logcli client already uses batches to fetch results. We could change it so that it returns an iterator over the entries. This way we won't fetch too much all at once.

`			resultLength, lastEntry = q.printResult(resp.Data.Result, out, lastEntry)
`

Thanks a lot, this streaming log note is really great.👍👍👍

I took a look at the implementation of logcli. The advantage of streaming is that it can continuously output the results to the cli command line.

But our remote read this time is to aggregate the results of multiple loki clusters, and then summarize the results uniformly, and return them to grafana at one time. At present, the http request of grafana log explore cannot implement streaming to print log plaintext.

Therefore, the implementation of remote read in this PR can not introduce streaming read for the time being. This will add a lot of code complexity, and grafana log explore cannot enjoy the same experience improvement as logcli.

Maybe we can complete this streaming read in the next PR, I am willing to go to grafana/grafana's loki datasource code to complete streaming printing from the ui source to the loki server end to end.  

In fact, this loki log explore supports streaming is also the result that our team has always wanted. The log explore of newRelic has a similar feature.

Reply:
I didn't mean to implement streaming. I meant to fetch the results from the remote in batches like `DoQuery` does and return and iterator instead. My assumption was that we are not aggregating everything because of the limit. I need to take a closer look if we could do batching for metrics as well.

Reply:
👍 OK，I got it, I will try to implement SelectLogs+batch next week.